File: app/src/main/java/com/pgssoft/mvvm/viewmodels/MainActivityViewModel.java
package com.pgssoft.mvvm.viewmodels;

import com.pgssoft.mvvm.model.api.ApiTable;
import com.pgssoft.mvvm.model.database.Rate;
import com.pgssoft.mvvm.model.database.Table;
import com.pgssoft.mvvm.services.interfaces.ApiService;
import com.pgssoft.mvvm.services.interfaces.ApiServiceCallback;
import com.pgssoft.mvvm.services.interfaces.MapperService;
import com.pgssoft.mvvm.services.interfaces.Repository;
import com.pgssoft.mvvm.viewmodels.interfaces.IMainActivityAccess;
import com.pgssoft.mvvm.views.adapters.interfaces.handlers.RateItemHandler;
import com.pgssoft.mvvm.views.adapters.interfaces.providers.BaseItemProvider;

import java.util.List;

/**
 * Created by bstokrocki on 29.01.2017.
 */
public class MainActivityViewModel implements BaseItemProvider<Rate>, RateItemHandler {
    private final IMainActivityAccess activityAccess;
    private final ApiService apiService;
    private final Repository repository;

    private List<Rate> rates;

    public MainActivityViewModel(IMainActivityAccess activityAccess, ApiService apiService,
                                 Repository repository) {
        this.activityAccess = activityAccess;
        this.apiService = apiService;
        this.repository = repository;
    }

    public void onInfrastructureReady() {
        loadRates();
    }

    private void loadRates() {
        activityAccess.showProgressIndication(false);
        apiService.loadRates(new ApiServiceCallback<ApiTable[]>() {
            @Override
            public void onSuccess(ApiTable[] apiTables) {
                repository.saveTables(apiTables);
                rates = repository.getAllRates();

                activityAccess.displayRates(MainActivityViewModel.this, MainActivityViewModel.this);
                activityAccess.hideProgressIndication();
            }

            @Override
            public void onFailure(Throwable t) {
                activityAccess.hideProgressIndication();
            }
        });
    }

    @Override
    public List<Rate> getItems() {
        return rates;
    }

    @Override
    public void showRateDetails(Rate rate) {
        activityAccess.openRateDetailsScreen(rate);
    }
}

File: kotlin/app/src/main/java/com/pgssoft/kotlinplayground/viewmodel/MainActivityViewModel.kt
package com.pgssoft.kotlinplayground.viewmodel

import com.pgssoft.kotlinplayground.model.api.ApiTable
import com.pgssoft.kotlinplayground.model.db.Rate
import com.pgssoft.kotlinplayground.service.iface.ApiService
import com.pgssoft.kotlinplayground.service.iface.ApiServiceCallback
import com.pgssoft.kotlinplayground.service.iface.Repository
import com.pgssoft.kotlinplayground.view.adapter.handlers.RateItemHandler
import com.pgssoft.kotlinplayground.view.adapter.providers.BaseItemProvider
import com.pgssoft.kotlinplayground.viewmodel.iface.IMainActivityAccess

/**
 * Created by bstokrocki on 29.01.2017.
 */
class MainActivityViewModel(
        private val activityAccess: IMainActivityAccess,
        private val apiService: ApiService,
        private val repository: Repository) : BaseActivityViewModel(), BaseItemProvider<Rate>, RateItemHandler {

    override var items: List<Rate> = ArrayList()
    internal var allItems: List<Rate> = ArrayList()

    val logoUrl = "https://raw.githubusercontent.com/PGSSoft/AndroidMVVM/master/app/src/main/res/drawable-xxhdpi/pgs.png"

    fun onInfrastructureReady() {
        loadRates()
    }

    internal fun loadRates() {
        activityAccess.showProgressIndication(false)
        apiService.loadRates(object : ApiServiceCallback<Array<ApiTable>> {
            override fun onSuccess(apiTables: Array<ApiTable>) {
                repository.saveTables(apiTables)
                allItems = repository.getAllRates()
                clearFiltering()
                activityAccess.hideProgressIndication()
            }

            override fun onFailure(t: Throwable) {
                activityAccess.hideProgressIndication()
            }
        })
    }

    fun clearFiltering() {
        items = allItems
        activityAccess.displayRates(this@MainActivityViewModel)
    }

    fun filter(query: String) {
        items = allItems.filter { it.currencyName.contains(query, true) || it.currencyCode.contains(query, true) }
        activityAccess.displayRates(this@MainActivityViewModel)
    }

    override fun showRateDetails(rate: Rate) {
        activityAccess.openRateDetailsScreen(rate)
    }

    override fun onStart() {

    }

    override fun onStop() {

    }
}
